pkg/kubelet/cm/deviceplugin/device_plugin_handler.go

```diff
@@ -205,74 +406,73 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
 		// requests may fail if we serve them in mixed order.
 		// TODO: may revisit this part later if we see inefficient resource allocation
 		// in real use as the result of this.
-		resp, err := h.devicePluginManager.Allocate(resource, append(devices.UnsortedList(), allocated...))
+		resp, err := h.devicePluginManager.Allocate(resource, allocDevices.UnsortedList())
 		if err != nil {
-			return nil, err
+			return err
+		}
+
+		// Update internal cached podDevices state.
+		h.Lock()
+		if oldDevices := h.podDevices.containerDevices(podUID, contName, resource); oldDevices != nil {
+			freeDevices := oldDevices.Difference(allocDevices)
```

Comment:
I might be understanding this wrong but this part of the code seems to suggest that you might not get the same devices between reboots right ?

Wouldn't we want a container to keep the exact same devices between reboots ?

Reply:
Good point. I added this part when I thought we need to support resource downsize and upsize, but then learned from Vish that a pod's resource requirements can't change as of today. Simplified this part.